### DISCUSSION TITLE
Bond Module

### DISCUSSION THREAD

**Discussion by robert-zaremba** [regen-network/regen-ledger#815] [Ideas]
Technical discussion about Bond Module design.
https://github.com/regen-network/regen-ledger/issues/726

**Comment by robert-zaremba** [comment]
I'm reviewing the initial proto definition for bonds. It looks like a Bond here is an NFT? It will have a face value, but bonds won't be fungible nor divisible, right?
In that case I would strongly consider implementing bonds on top of Cosmos SDK NFT module.

**Reply by aaronc** [reply]
To me it actually strikes me as fungible because it seems like it can be split between different owners. @tonda100 @horavamartin can you clarify as to whether or not it is divisible?

**Reply by robert-zaremba** [reply]
This what I would think about, but the current design has a single bond (with unique ID -> https://github.com/regen-network/regen-ledger/issues/726#issuecomment-1048768152) with a face value.... so a followup thread / question below.

**Comment by robert-zaremba** [comment]
### Bond Emission.

1. Normally, whenever bonds are emitted, there is a stock of bonds and each bond has a face value. That bonds are fungible. This design doesn't imply that - I see that emission creates a new denom with face value, but doesn't define amount of bonds (coins) to emit.
Moreover who and how should decide about the face value currency? Is it going to be denominated in Regen?
2. Are there multiple emissions for the same Bond supported?

**Reply by yddgd** [reply]
@robert-zaremba , I am not sure whether we share the same understanding of how the bonds work.

There are the expectations we have from the bonds..

1. The corporate bonds are generally issued by the issuer (borrower), sold to the underwriter, who then sells the bonds to the bondholders (lenders).
2. One issuer can issue (emit) more than one issue (emission).
3. The emission of the bond is the whole debt in the face currency. In our case preferably a stablecoin.
4. It can be divided to the degree chosen by the issuer and can be the whole issue or minimal possible denomination of the currency or anywhere between. In our case preferably the minimal currency denomination.
   - This equals its face value - the amount returned by the borrower to the bondholder at the maturity.
5. The reward for the debt is coupon or coupon rate (also called interest rate), which is set at issuance and fixed over the whole life of the bond.
   - It is paid out usually annually.
   - The frequency can be also semi-annually, quarterly, monthly or even daily.
   - Coupon rate is the annual percentage from the face value, coupon is the percentage expressed in absolute amount but divided by frequency.
6. The entity with the minimal face value is (single) bond.
   - For our case it can be useful to call "bonds" also the multiples of the single bond.
7. The bonds within one emission are fungible (they have the same maturity and coupon rate).
   - The bonds are not fungible among emission, even from the same issuer.
8. In case the bond is negotiable (can be re-sold - e.g. on the secondary market) and bondholder owns at least single bond, the bond can be sold or partially sold.
   - In case of a full sale, the ownership of the original bondholder ownership ceases to exist and the new ownership is created.
   - In case of a partial sale, the original ownership is updated (lowered by the single bonds multiple sold) and the new ownership is created.
9. The sale can be carried out in any currency, but the original currency makes the most sense.
10. At maturity, the bondholder is paid back the face value of the bond(s) they own and also usually the last coupon.
11. The ownership at maturity does not change nor ceases.
  

Is this something we understand in the same way?

**Reply by robert-zaremba** [reply]
I was writing once a bond contract for one client. In that project we had:
* predefined emissions: when a new emission is created, all bonds in that emissions were created and assigned to the issuer.
* at maturity the bond holder can redeem the bond for the principle + unpaid coupons. (PS: handling coupons will require a separate discussion - we need to remember if a coupon payment for a given bond was done or not).
* Bonds in given emission were fungible, however for handling coupons I designed a multi-level bond - each bond has sub denoms representing a stage with respect to coupons paid.

> 10. At maturity, the bondholder is paid back the face value of the bond(s) they own and also usually the last coupon.
> 11. The ownership at maturity does not change nor ceases.

It is not clear to me why you propose to not cease the bond at maturity? I think it should function as I wrote in the second bullet point.

**Reply by robert-zaremba** [reply]
Currently, in your [`MsgIssueBond`](https://github.com/regen-network/regen-ledger/pull/801/files#diff-16888172986b0ed586bf547426943313970acce2be2f20d7bb5567bcc7f55c4fR10) proto, it's not clear to me if this will issue a single bond or perform a whole emission.

**Reply by yddgd** [reply]
@robert-zaremba , you are right, since it is the batch issuance, the better name would be probably `MsgIssueBondBatch`. @horavamartin, please update.

**Reply by yddgd** [reply]
> predefined emissions: when a new emission is created, all bonds in that emissions were created and assigned to the issuer.

Yes, they might be assigned to the issuer as well, although I would prefer to be assigned directly to the owner.

> at maturity the bond holder can redeem the bond for the principle + unpaid coupons. (PS: handling coupons will require a separate discussion - we need to remember if a coupon payment for a given bond was done or not).

> It is not clear to me why you propose to not cease the bond at maturity? I think it should function as I wrote in the second bullet point.

In my opinion, there should not be any bondholder action required for either coupon or face value payout on coupon payout term or maturity. Thats why I would prefer daily coupon, but it seems that this option is undesirable. Hence we indeed have to keep track of the paid out coupons. What would be your suggestions to do so?

At 10. and 11., I probably not expressed that correctly. What should happen is setting the bond the status matured, but the ownership does not change, similarly to the retired ecocredit.

> Bonds in given emission were fungible, however for handling coupons I designed a multi-level bond - each bond has sub denoms representing a stage with respect to coupons paid.

You probably need to take me further with the multi-level bond - is there somewhere for me to read more to understand that better?

**Reply by robert-zaremba** [reply]
> Yes, they might be assigned to the issuer as well, although I would prefer to be assigned directly to the owner.

Issuer is the owner, right? And then he will sell the bonds.

**Reply by robert-zaremba** [reply]
> In my opinion, there should not be any bondholder action required for either coupon or face value payout on coupon payout term or maturity. Thats why I would prefer daily coupon, but it seems that this option is undesirable. Hence we indeed have to keep track of the paid out coupons. What would be your suggestions to do so?

I would prefer to limit the "automatic actions" handled by a blockchain. Especially something that is repeated. It would require end blocker function and could be expensive if there will be many many bonds. 

Instead, we should use an algorithm which will accumulate the unpaid rewards, and user will be able to redeem it. It would require break down the bond into sub-denoms, to know if the given bond received reward or not.

**Reply by robert-zaremba** [reply]
> You probably need to take me further with the multi-level bond - is there somewhere for me to read more to understand that better?

The idea is that we need to distinguish bonds which received payments. So we need to break down the time into epochs according to the reward payment. For bond `B`, we will have `B_1, B_2, .... B_n`. 
At epoch `e2`, if a user has a x bonds from epoch `e1`: `B_e1` he can exchange it to x bonds `B_e1` and `x * (e2-e1)` reward units.

**Comment by neuschool** [comment]
I fully agree that 'Green Bonds' are a significant driver in the adoption of web3 for the carbon vertical. I would like to suggest adding layers of MRV / remote sensing would be a big value add. Currently bonds are very un-transparent in terms of fulfilling the conditions of the bond agreement. 

With web3 and remote sensing tech we can ensure that the conditions are met and in addition stream the earnings / coupons into the market on a more frequent basis creating a more consistent and predictable cash flow model. This all supports a more efficient market and de-risks the bond as the buyer gets more frequent coupon payments...
